src/nvim/eval.c

```diff
@@ -18151,19 +18176,25 @@ static void f_writefile(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     }
   }
 
-  // Always open the file in binary mode, library functions have a mind of
-  // their own about CR-LF conversion.
-  char_u *fname = get_tv_string(&argvars[1]);
-  FILE *fd;
-  if (*fname == NUL || (fd = mch_fopen((char *)fname,
-       append ? APPENDBIN : WRITEBIN)) == NULL) {
-    EMSG2(_(e_notcreate), *fname == NUL ? (char_u *)_("<empty>") : fname);
-    rettv->vval.v_number = -1;
+  const char *const fname = (const char *)get_tv_string(&argvars[1]);
+  FileDescriptor *fp;
+  int error;
+  rettv->vval.v_number = -1;
+  if (*fname == NUL) {
+    EMSG(_("E482: Can't open file with an empty name"));
+  } else if ((fp = file_open_new(&error, fname,
+                                 ((append ? kFileAppend : kFileTruncate)
+                                  | kFileCreate), 0666)) == NULL) {
+    emsgf(_("E482: Can't open file %s for writing: %s"),
+          fname, os_strerror(error));
   } else {
-    if (write_list(fd, argvars[0].vval.v_list, binary) == false) {
-      rettv->vval.v_number = -1;
+    if (write_list(fp, argvars[0].vval.v_list, binary)) {
+      rettv->vval.v_number = 0;
+    }
+    if ((error = file_free(fp)) != 0) {
+      emsgf(_("E80: Error when closing file %s: %s"),
+            fname, os_strerror(error));
```

Comment:
The use of `fname` here looks like UB, because `get_tv_string()` uses a static buffer for Number variables.
`fname` is defined as a `const char *`, but the buffer content could be changed by the previous call to  `write_list()` if the list contains a Number variable.